src/Control/Monad/Free.purs

```diff
@@ -132,52 +82,63 @@ bindFC fa t = bindF fa (liftCoyonedaTF t)
+foldFree k = tailRecM go
   where
-  resumeGosub :: Free f a -> Either (f (Free f a)) (Free f a)
-  resumeGosub (Gosub g) =
-    runExists (\(GosubF v) -> case v.a unit of
-                                   Pure a -> Right (v.f a)
-                                   Free t -> Left ((\h -> h >>= v.f) <$> t)
-                                   Gosub h -> runExists (\(GosubF w) -> Right (w.a unit >>= (\z -> w.f z >>= v.f))) h) g
-
--- | `runFree` runs a computation of type `Free f a`, using a function which unwraps a single layer of
--- | the functor `f` at a time.
+  go :: Free f a -> m (Either (Free f a) a)
+  go f =
+    case toView f of
+         Return a -> Right <$> pure a
+         Bind g i -> (Left <<< i) <$> k g
+
+-- | Run a free monad with a function that unwraps a single layer of the functor `f` at a time,
 runFree :: forall f a. (Functor f) => (f (Free f a) -> Free f a) -> Free f a -> a
-runFree fn = runIdentity <<< runFreeM (Identity <<< fn)
+runFree k = runIdentity <<< runFreeM (Identity <<< k)
 
--- | `runFreeM` runs a compuation of type `Free f a` in any `Monad` which supports tail recursion.
+-- | Run a free monad with a function mapping a functor `f` to a tail-recursive monad `m`.
 -- | See the `MonadRec` type class for more details.
 runFreeM :: forall f m a. (Functor f, MonadRec m) => (f (Free f a) -> m (Free f a)) -> Free f a -> m a
-runFreeM fn = tailRecM \f ->
-  case resume f of
-    Left fs -> Left <$> fn fs
-    Right a -> return (Right a)
-
--- | `runFreeC` is the equivalent of `runFree` for type constructors transformed with `Coyoneda`,
--- | hence we have no requirement that `f` be a `Functor`.
-runFreeC :: forall f a. (forall a. f a -> a) -> FreeC f a -> a
-runFreeC nat = runIdentity <<< runFreeCM (Identity <<< nat)
-
--- | `runFreeCM` is the equivalent of `runFreeM` for type constructors transformed with `Coyoneda`,
--- | hence we have no requirement that `f` be a `Functor`.
-runFreeCM :: forall f m a. (MonadRec m) => Natural f m -> FreeC f a -> m a
-runFreeCM nat = runFreeM (liftCoyonedaTF nat)
+runFreeM k = tailRecM go
+  where
+  go :: Free f a -> m (Either (Free f a) a)
+  go f =
+    case toView f of
+         Return a -> Right <$> pure a
+         Bind g i -> Left <$> k (i <$> g)
+
+fromView :: forall f a. FreeView f a Val -> Free f a
+fromView f = Free (unsafeCoerceFreeView f) empty
+
+toView :: forall f a. Free f a -> FreeView f a Val
+toView (Free v s) =
+  case v of
+       Return a -> case (uncons s) of
+                        Nothing -> Return (unsafeCoerceVal a)
+                        Just (Tuple h t) -> toView (unsafeCoerceFree (concatF ((runExpF h) a) t))
+       Bind f k -> Bind f (\a -> unsafeCoerceFree (concatF (k a) s))
+  where
+  concatF :: Free f Val -> CatList (ExpF f) -> Free f Val
+  concatF (Free v l) r = Free v (l <> r)
+
+  runExpF :: forall f. ExpF f -> (Val -> Free f Val)
+  runExpF (ExpF k) = k
+
+unsafeCoerceVal :: forall a. Val -> a
```

Comment:
I'd be up to move all the unsafe coerce functions into appropriate where blocks. I think it would make the code a bit cleaner.